File: dstar_lite.py
import heapq

class DStarLite:
    """
    A minimal implementation of D* Lite for finding the shortest path dynamically.
    D* Lite searches backwards from the Goal to the Start, which makes it extremely
    efficient at replanning when edge costs near the Start node change.
    """
    def __init__(self, graph, start, goal):
        self.graph = graph
        self.start = start
        self.goal = goal
        
        # Priority queue
        self.U = [] 
        self.km = 0
        
        # rhs (one step lookahead) and g (objective distance) values
        self.rhs = {}
        self.g = {}
        
        # Initialize all nodes to infinity
        for node in self.graph.edges:
            self.rhs[node] = float('inf')
            self.g[node] = float('inf')
            
        # The goal is our starting point for the backward search
        self.rhs[self.goal] = 0.0
        heapq.heappush(self.U, (self.calculate_key(self.goal), self.goal))
        
    def heuristic(self, a, b):
        # Without lat/lon coordinates to calculate a real distance heuristic, 
        # we return 0. This effectively makes it behave like dynamic Dijkstra's.
        return 0.0
        
    def calculate_key(self, s):
        k1 = min(self.g[s], self.rhs[s]) + self.heuristic(self.start, s) + self.km
        k2 = min(self.g[s], self.rhs[s])
        return (k1, k2)
        
    def update_vertex(self, u):
        if u != self.goal:
            # rhs[u] = min over all neighbors s' of (cost(u, s') + g(s'))
            min_val = float('inf')
            for neighbor, cost in self.graph.get_neighbors(u).items():
                val = cost + self.g[neighbor]
                if val < min_val:
                    min_val = val
            self.rhs[u] = min_val
            
        # Remove u from U if it's in there (this is inefficient for huge graphs, 
        # but perfectly fine for MST/Barangay scale graphs)
        self.U = [item for item in self.U if item[1] != u]
        heapq.heapify(self.U)
        
        # If the vertex is inconsistent, put it back on the queue
        if self.g[u] != self.rhs[u]:
            heapq.heappush(self.U, (self.calculate_key(u), u))
            
    def compute_shortest_path(self):
        while len(self.U) > 0:
            self.U.sort()
            u_key, u = self.U[0]
            
            # If the start node is consistent and has the lowest key, we are done
            if u_key >= self.calculate_key(self.start) and self.rhs[self.start] == self.g[self.start]:
                break
                
            heapq.heappop(self.U)
            
            k_old = u_key
            k_new = self.calculate_key(u)
            
            if k_old < k_new:
                heapq.heappush(self.U, (k_new, u))
            elif self.g[u] > self.rhs[u]:
                # Vertex is over-consistent (found a shorter path)
                self.g[u] = self.rhs[u]
                for neighbor in self.graph.get_neighbors(u):
                    self.update_vertex(neighbor)
            else:
                # Vertex is under-consistent (a path got longer)
                self.g[u] = float('inf')
                self.update_vertex(u)
                for neighbor in self.graph.get_neighbors(u):
                    self.update_vertex(neighbor)
                    
    def get_path(self):
        """Extracts the shortest path from start to goal based on the calculated g values."""
        if self.g[self.start] == float('inf'):
            return None # No path exists
            
        path = [self.start]
        current = self.start
        
        while current != self.goal:
            min_cost = float('inf')
            next_node = None
            
            for neighbor, cost in self.graph.get_neighbors(current).items():
                val = cost + self.g[neighbor]
                if val < min_cost:
                    min_cost = val
                    next_node = neighbor
                    
            if next_node is None:
                break
                
            path.append(next_node)
            current = next_node
            
        return path
        
    def dynamically_update_edge(self, u, v, new_cost):
        """
        The power of D* Lite: Call this when a road condition changes.
        It instantly updates the graph and recalculates the new best path.
        """
        self.graph.update_edge_cost(u, v, new_cost)
        
        self.update_vertex(u)
        self.update_vertex(v)
        
        self.compute_shortest_path()

File: static/js/dstar_lite.js
class PriorityQueue {
    constructor() {
        this.heap = [];
    }
    
    _parent(i) { return Math.floor((i - 1) / 2); }
    _left(i) { return 2 * i + 1; }
    _right(i) { return 2 * i + 2; }
    
    _compare(a, b) {
        if (a.key[0] < b.key[0]) return -1;
        if (a.key[0] > b.key[0]) return 1;
        if (a.key[1] < b.key[1]) return -1;
        if (a.key[1] > b.key[1]) return 1;
        return 0;
    }
    
    _swap(i, j) {
        const temp = this.heap[i];
        this.heap[i] = this.heap[j];
        this.heap[j] = temp;
    }
    
    push(key, node) {
        this.heap.push({ key, node });
        this._siftUp(this.heap.length - 1);
    }
    
    pop() {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop();
        
        const top = this.heap[0];
        this.heap[0] = this.heap.pop();
        this._siftDown(0);
        return top;
    }
    
    peek() {
        return this.heap.length > 0 ? this.heap[0] : null;
    }
    
    remove(node) {
        const idx = this.heap.findIndex(item => item.node === node);
        if (idx !== -1) {
            this.heap.splice(idx, 1);
            // Re-heapify simply by sorting (inefficient but perfectly fine for our scale)
            this.heap.sort(this._compare);
        }
    }
    
    _siftUp(i) {
        while (i > 0 && this._compare(this.heap[i], this.heap[this._parent(i)]) < 0) {
            this._swap(i, this._parent(i));
            i = this._parent(i);
        }
    }
    
    _siftDown(i) {
        let minIndex = i;
        const l = this._left(i);
        if (l < this.heap.length && this._compare(this.heap[l], this.heap[minIndex]) < 0) {
            minIndex = l;
        }
        const r = this._right(i);
        if (r < this.heap.length && this._compare(this.heap[r], this.heap[minIndex]) < 0) {
            minIndex = r;
        }
        if (i !== minIndex) {
            this._swap(i, minIndex);
            this._siftDown(minIndex);
        }
    }

    get length() {
        return this.heap.length;
    }
}

class DStarLite {
    constructor(graph, start, goal) {
        this.graph = graph;
        this.start = start;
        this.goal = goal;
        
        this.U = new PriorityQueue();
        this.km = 0;
        
        this.rhs = {};
        this.g = {};
        
        for (let node in this.graph.edges) {
            this.rhs[node] = Infinity;
            this.g[node] = Infinity;
        }
        
        this.rhs[this.goal] = 0.0;
        this.U.push(this.calculate_key(this.goal), this.goal);
    }
    
    heuristic(a, b) {
        return 0.0;
    }
    
    calculate_key(s) {
        let k1 = Math.min(this.g[s], this.rhs[s]) + this.heuristic(this.start, s) + this.km;
        let k2 = Math.min(this.g[s], this.rhs[s]);
        return [k1, k2];
    }
    
    update_vertex(u) {
        if (u !== this.goal) {
            let min_val = Infinity;
            const neighbors = this.graph.get_neighbors(u);
            for (let neighbor in neighbors) {
                let cost = neighbors[neighbor];
                let val = cost + this.g[neighbor];
                if (val < min_val) {
                    min_val = val;
                }
            }
            this.rhs[u] = min_val;
        }
        
        this.U.remove(u);
        
        if (this.g[u] !== this.rhs[u]) {
            this.U.push(this.calculate_key(u), u);
        }
    }
    
    compute_shortest_path() {
        while (this.U.length > 0) {
            let top = this.U.peek();
            let u_key = top.key;
            let u = top.node;
            
            let start_key = this.calculate_key(this.start);
            
            // if (u_key >= start_key)
            if (this.U._compare({key: u_key}, {key: start_key}) >= 0 && this.rhs[this.start] === this.g[this.start]) {
                break;
            }
            
            this.U.pop();
            
            let k_old = u_key;
            let k_new = this.calculate_key(u);
            
            if (this.U._compare({key: k_old}, {key: k_new}) < 0) {
                this.U.push(k_new, u);
            } else if (this.g[u] > this.rhs[u]) {
                this.g[u] = this.rhs[u];
                const neighbors = this.graph.get_neighbors(u);
                for (let neighbor in neighbors) {
                    this.update_vertex(neighbor);
                }
            } else {
                this.g[u] = Infinity;
                this.update_vertex(u);
                const neighbors = this.graph.get_neighbors(u);
                for (let neighbor in neighbors) {
                    this.update_vertex(neighbor);
                }
            }
        }
    }
    
    get_path() {
        if (this.g[this.start] === Infinity) {
            return null;
        }
        
        let path = [this.start];
        let current = this.start;
        
        while (current !== this.goal) {
            let min_cost = Infinity;
            let next_node = null;
            
            const neighbors = this.graph.get_neighbors(current);
            for (let neighbor in neighbors) {
                let cost = neighbors[neighbor];
                let neighbor_cost = cost + this.g[neighbor];
                
                // Prioritize natural roads over V- connections in case of a tie
                if (Math.abs(neighbor_cost - min_cost) < 1e-6) {
                    if (next_node && next_node.startsWith('V-') && !neighbor.startsWith('V-')) {
                        next_node = neighbor;
                    }
                } else if (neighbor_cost < min_cost) {
                    min_cost = neighbor_cost;
                    next_node = neighbor;
                }
            }
            
            if (!next_node) {
                return null;
            }
            
            path.push(next_node);
            current = next_node;
        }
        return path;
    }
}
